### PR TITLE
feat!: Track circuit extensions and read/write packages

### PR DESCRIPTION
**Pull request by aborgna-q**
Removes the guppy-specific and adds supports for loading functions packages and standalone hugrs.

Temporarily keeps track of the required extensions for the hugr in an optional `Circuit::required_extensions` field until https://github.com/CQCL/hugr/issues/1613 gets implemented. Fallbacks to a default set when loading bare hugrs.

Note that storing a circuit with a non-root parent is currently an error. We'll need to store some pointer to the entrypoint on the hugr's metadata, and that'll require some serialization-stable path encoding. I'll open an issue for that .

blocked-by: https://github.com/CQCL/hugr/pull/1621. I'll remove the patch in cargo.toml once that gets released.

drive-by: Use `circuit_hash` for the `PartialEq` implementation of circuits. The derived equality failed on graphs with different node indices.

BREAKING CHANGE: Removed `load_guppy_*` methods. Use `Circuit::load_function_reader` instead.

### CODE REVIEW

File: tket2-py/tket2/circuit/build.py [hunk 1]
```diff
@@ -20,17 +21,33 @@ class CircBuild(TrackedDfg):
     def with_nqb(cls, n_qb: int) -> CircBuild:
         return cls(*[tys.Qubit] * n_qb, track_inputs=True)
 
+    def finish_hugr(self) -> Hugr:
+        """Finish building the package by setting all the qubits as the output
+        and wrap it in a hugr package with the required extensions.
```
Comment (ss2165) [inline tket2-py/tket2/circuit/build.py:26]: doesn't appear to be wrapping in a package?

File: tket2-py/tket2/circuit/build.py [hunk 2]
```diff
@@ -42,13 +59,26 @@ def finish_package(
             *(other_extensions or []),
         ]
 
-        return Package(modules=[self.hugr], extensions=extensions)
+        # Convert the DFG into a Function definition
+        dfg_op = self.hugr[self.hugr.root].op
+        assert type(dfg_op) is ops.DFG, "CircBuild must have a Dfg root"
+        self.hugr[self.hugr.root].op = ops.FuncDefn(
+            function_name, inputs=dfg_op.inputs, _outputs=dfg_op.outputs
+        )
```
Comment (ss2165) [inline tket2-py/tket2/circuit/build.py:67]: hugr rewriting is so much easier in python 🤔

File: tket2/src/circuit.rs [hunk 3]
```diff
@@ -64,11 +88,22 @@ lazy_static! {
...
     };
+
+    /// A default set of required extensions for a circuit,
+    /// used when loading with hugr with no pre-defined extension set.
+    ///
+    /// We should be able to drop this once hugrs embed their required extensions.
+    /// See https://github.com/CQCL/hugr/issues/1613
+    static ref DEFAULT_REQUIRED_EXTENSIONS: Vec<Extension> = extension::REGISTRY.iter().map(|(_, ext)| ext.clone()).collect();
 }
+/// The [IGNORED_EXTENSION_OPS] definition depends on the buggy behaviour of [`NamedOp::name`], which returns bare names instead of scoped names on some cases.
```
Comment (ss2165) [inline tket2/src/circuit.rs:99]: nice comment 🙏

File: tket2/src/serialize.rs [hunk 4]
```diff
@@ -1,13 +1,421 @@
...
+        // TODO: Read the entrypoint from the metadata.
+        let root = hugr.root();
+        Ok(Circuit::try_new(hugr, root)?)
+    }
+}
+
+/// Error type for conversion between `Op` and `OpType`.
+#[derive(Debug, Display, Error, From)]
+#[non_exhaustive]
+pub enum CircuitStoreError {
```
Comment (ss2165) [inline tket2/src/serialize.rs:135]: when are we dropping the word "Circuit" 👀

File: tket2/src/serialize.rs [hunk 5]
```diff
@@ -1,13 +1,421 @@
...
+    PackageStore(PackageError),
+    /// The circuit's parent is not the root of the HUGR.
+    #[display("The circuit's parent {parent} is not the root of the HUGR.")]
+    NonRootCircuit {
+        /// The parent node.
+        parent: Node,
+    },
+}
+
+/// Error type for conversion between `Op` and `OpType`.
```
Comment (ss2165) [inline tket2/src/serialize.rs:151]: maybe add tket2/tket1 prefixes to Op and OpType to distinguish

File: tket2/src/serialize.rs [hunk 6]
```diff
@@ -1,13 +1,421 @@
...
+fn find_function(hugr: Hugr, function_name: &str) -> Result<Circuit, CircuitLoadError> {
+    // Find the root module.
+    let module = hugr.root();
+    if !OpTag::ModuleRoot.is_superset(hugr.get_optype(module).tag()) {
+        return Err(CircuitLoadError::NonModuleRoot {
+            root_op: hugr.get_optype(module).clone(),
+        });
+    }
+
+    // Find the function declaration.
```
Comment (ss2165) [inline tket2/src/serialize.rs:280]:
```suggestion
    // Find the function definition.
```

File: tket2/src/serialize.rs [hunk 7]
```diff
@@ -1,13 +1,421 @@
...
+
+        let mut mod1 = ModuleBuilder::new();
+        define("apple", &mut mod1);
+        define("banana", &mut mod1);
+        let mod1 = mod1.finish_prelude_hugr().unwrap();
+
+        let mut mod2 = ModuleBuilder::new();
+        define("foo", &mut mod2);
+        define("bar", &mut mod2);
+        define("banana", &mut mod2);
```
Comment (ss2165) [inline tket2/src/serialize.rs:366]: 🍌

File: tket2/src/serialize.rs [hunk 8]
```diff
@@ -1,13 +1,421 @@
...
+    fn func_circuit_store(function_circ: Circuit) {
+        let mut buf = Vec::new();
+        function_circ.to_package_writer(&mut buf).unwrap();
+        let circ2 = Circuit::load_function_reader(Cursor::new(buf), "banana").unwrap();
+
+        assert_eq!(function_circ, circ2);
+    }
+
+    #[rstest]
+    fn serialize_errors(multi_module_pkg: Package) {
```
Comment (ss2165) [inline tket2/src/serialize.rs:398]: there's still a few errors missing coverage, up to you about adding more tests

**Review by ss2165 — COMMENTED** [review]
🍌

File: tket2/src/serialize.rs (hunk 4 above)
Comment (aborgna-q) [reply tket2/src/serialize.rs:135]: Soon™️
(it should be my next PR)

**Review by ss2165 — APPROVED** [review]
🍌

File: tket2/src/serialize/guppy.rs [hunk 9]
```diff
@@ -1,162 +0,0 @@
...
-///
-/// Currently we only support functions where the CFG operation has a single `DataflowBlock` child,
-/// which we use as the root of the circuit. We (currently) do not support control flow primitives.
-///
-/// # Errors
-///
-/// - If the root of the HUGR is not a module operation.
-/// - If the function is not found in the module.
-/// - If the function has control flow primitives.
-pub fn find_function(hugr: Hugr, function_name: &str) -> Result<Circuit, CircuitLoadError> {
```
Comment (aborgna-q) [inline tket2/src/serialize/guppy.rs:58]: This was required to unwrap the CFGs generated by guppy in all its function definitions. See #777

**Pull request merged**